### DISCUSSION TITLE
How can I use a generated client with @Autowired for Spring

### DISCUSSION THREAD

**Discussion by xetra11** [ExpediaGroup/graphql-kotlin#908] [Q&A] [answered]
I am using the gradle plugin now and successfully created a client for my GraphQL query. Now I want to use this via Dependency Injection by Spring.

I saw that the generated class is not configured as Spring Component and wonder how I can achieve this to easily use the client in my spring services.

Here is my configuration:
```
plugins: {
      id("com.expediagroup.graphql") version "4.0.0-alpha.6"
}
implementation("com.expediagroup:graphql-kotlin-spring-client:4.0.0-alpha.6")
graphql {
    client {
        allowDeprecatedFields = true
        clientType = GraphQLClientType.WEBCLIENT
        packageName = "com.company.network.graphql"
        endpoint = "http://myintro/graphql"
    }
}
```
I wonder if I need to add something to let it generate and apply it to a spring context

**Comment by dariuszkuc** [comment, marked as answer]
In order to use the generated clients as Spring beans you will need to manually register them with Spring context, e.g. given `HelloWorldQuery` you could create it as

```kotlin
@Configuration
class QueryConfiguration {
    @Bean
    fun graphQLWebClient(builder: WebClient.Builder) = 
        GraphQLWebClient(url = "http://localhost:8080/graphql", builder = builder)
    
    @Bean
    fun helloWorldQuery(graphQLWebClient: GraphQLWebClient) = HelloWorldQuery(graphQLWebClient)
}
```

You can then autowire `helloWorldQuery: HelloWorldQuery` bean to any of your services.

**Comment by dariuszkuc** [comment]
@smyrick wondering whether we should automatically add `@Component` annotation to the webclient queries so they would be automatically picked up by Spring. 

Additional considerations - `GraphQLWebClient` would still have to be manually registered with Spring context and it is possible to have different queries pointed at different services so automatic component scan would not work if there are multiple web clients.

**Reply by xetra11** [reply]
I was thinking about the same - forgot to suggest it. You can use qualifiers to tell them apart from usual spring beans maybe?

**Reply by smyrick** [reply]
We can add a component annotation. If we add a `@ConditionalOnMissingBean` on components will that allow anyone to override the client as well?

**Reply by dariuszkuc** [reply]
We won't be able to generically create the graphQL web client bean unless we provide some extra logic on top of it (at a minimum we need target URL) -> this could morph into another autoconfiguration which sounds like an overkill. With multiple web clients available you would need to use their names/qualifiers to distinguish them which adds additional complexities on how to determine those while generating the client code.

So while we could add `@Component` on the generated queries, due to the above I don't think it makes sense to complicate the generator logic that would only work in subset of scenarios.
